src/containers.py

```diff
@@ -0,0 +1,186 @@
+    prepared_path = orig_path.as_posix()
+    # Check whether the path matches an absolute Windows path with a drive letter
+    match = re.match(r'^([A-Z]:)(.*)', prepared_path)
+    if match:
+        # The first group is the drive such as C:
+        drive = match.group(1).lower()[0]
+        # The second group is the rest of the path such as /Users/me
+        prepared_path = match.group(2)
+        prepared_path = '//' + drive + prepared_path
+    return prepared_path
+
+
+def convert_docker_path(src_path: PurePath, dest_path: PurePath, file_path: Union[str, PurePath]) -> PurePosixPath:
+    """
+    Convert a file_path that is in src_path to be in dest_path instead.
+    For example, convert /usr/mydir and /usr/mydir/myfile and /tmp to /tmp/myfile
+    @param src_path: source path that is a parent of file_path
+    @param dest_path: destination path
+    @param file_path: filename that is under the source path
+    @return: a new path with the filename relative to the destination path
+    """
+    rel_path = file_path.relative_to(src_path)
+    return PurePosixPath(dest_path, rel_path)
+
+
+# TODO consider a better default environment variable
+# Follow docker-py's naming conventions (https://docker-py.readthedocs.io/en/stable/containers.html)
+# Technically the argument is an image, not a container, but we use container here.
+def run_container(framework: str, container_suffix: str, command: List[str], volumes: List[Tuple[PurePath, PurePath]], working_dir: str, environment: str = 'SPRAS=True'):
+    """
+    Runs a command in the container using Singularity or Docker
+    @param framework: singularity or docker
+    @param container: name of the DockerHub container without the 'docker://' prefix
+    @param command: command to run in the container
+    @param volumes: a list of volumes to mount where each item is a (source, destination) tuple
+    @param working_dir: the working directory in the container
+    @param environment: environment variables to set in the container
+    @return: output from Singularity execute or Docker run
+    """
+    normalized_framework = framework.casefold()
+
+    container = config.config.container_prefix + "/" + container_suffix
+    if normalized_framework == 'docker':
+        return run_container_docker(container, command, volumes, working_dir, environment)
+    elif normalized_framework == 'singularity':
+        return run_container_singularity(container, command, volumes, working_dir, environment)
+    else:
+        raise ValueError(f'{config.config.framework} is not a recognized container framework. Choose "docker" or "singularity".')
+
+
+# TODO any issue with creating a new client each time inside this function?
+# TODO environment currently a single string (e.g. 'TMPDIR=/OmicsIntegrator1'), should it be a list?
+def run_container_docker(container: str, command: List[str], volumes: List[Tuple[PurePath, PurePath]], working_dir: str, environment: str = 'SPRAS=True'):
```

Comment:
Nope, nothing should have changed. In our last meeting, we discussed that container-related functions were getting nuanced enough that it would be a good idea to move them to their own `containers.py`. The only changes to any of the functions was in `run_container`, which now access configuration options to determine the container prefix, ie something like `docker.io/reedcompbio`

Do we think the container prefix is something that should be configurable per PRM? If so, I can move that up one level and pass `run_container` the entire name instead of parsing it out inside the function. I wasn't sure how much flexibility to shoot for at this point, since those types of features haven't been requested.